File: Backend/src/Iit/Iit.Fibertest.Graph/DomainEntities/Graph/Rtu.cs
using Iit.Fibertest.Dto;

namespace Iit.Fibertest.Graph
{
    [Serializable]
    public class Rtu
    {
        public Guid Id { get; set; }

        // public string OtauId { get; set; } // in VeEX RTU main OTAU has its own ID
        public string OtdrId { get; set; } = string.Empty; // ditto
        public VeexOtau MainVeexOtau { get; set; } = new(); // in Veex RTU it is a separate unit

        public RtuMaker RtuMaker { get; set; }
        public string Mfid { get; set; } = null!;
        public string Mfsn { get; set; } = null!;
        public string Omid { get; set; } = null!;
        public string Omsn { get; set; } = null!;

        public Guid NodeId { get; set; }
        public string Title { get; set; } = null!;
        public string Comment { get; set; } = string.Empty;

        public NetAddress MainChannel { get; set; } = new(@"", -1);
        public RtuPartState MainChannelState { get; set; }
        public NetAddress ReserveChannel { get; set; } = new("", -1);
        public RtuPartState ReserveChannelState { get; set; }
        public bool IsReserveChannelSet { get; set; }
        public NetAddress OtdrNetAddress { get; set; } = new(@"0.0.0.0", 1500); // real address
        public bool IsAvailable => MainChannelState == RtuPartState.Ok ||
                                   ReserveChannelState == RtuPartState.Ok;
        public bool IsAllRight => MainChannelState != RtuPartState.Broken &&
                                  ReserveChannelState != RtuPartState.Broken;

        // pair OTAU ID - is OK or not
        private Dictionary<Guid, bool> _otauStates = new();
        public void SetOtauState(Guid otauId, bool isOk)
        {
            // if (_otauStates == null)
            //     _otauStates = new Dictionary<Guid, bool>();
            _otauStates[otauId] = isOk;
        }

        public void RemoveOtauState(Guid otauId)
        {
            // if (_otauStates == null)
            //     _otauStates = new Dictionary<Guid, bool>();
            if (_otauStates.ContainsKey(otauId))
                _otauStates.Remove(otauId);
        }

        public RtuPartState BopState
        {
            get
            {
                // if (_otauStates == null)
                //     _otauStates = new Dictionary<Guid, bool>();
                return _otauStates.Count == 0
                    ? RtuPartState.NotSetYet
                    : _otauStates.Any(s => s.Value != true)
                        ? RtuPartState.Broken
                        : RtuPartState.Ok;
            }
        }
        
        public string Serial { get; set; } = string.Empty;
        public int OwnPortCount { get; set; }
        public bool IsInitialized => OwnPortCount != 0;
        public int FullPortCount { get; set; }

        public string PortCount => OwnPortCount == FullPortCount ? $@"{FullPortCount}" : $@"{OwnPortCount} / {FullPortCount}";

        public string Version { get; set; } = String.Empty;
        public string Version2 { get; set; } = string.Empty;

        public Dictionary<int, OtauDto> Children { get; set; } = new();

        public MonitoringState MonitoringState { get; set; }
        public Frequency PreciseMeas { get; set; } = Frequency.EveryHour;
        public Frequency PreciseSave { get; set; } = Frequency.DoNot;
        public Frequency FastSave { get; set; } = Frequency.DoNot;

        public TreeOfAcceptableMeasParams AcceptableMeasParams { get; set; } = new();

        public List<Guid> ZoneIds { get; set; } = new();

    }
}

File: WebFt30/src/app/core/store/models/ft30/rtu.ts
import { TreeOfAcceptableMeasurementParameters } from './acceptable-measurement-parameters';
import { Bop } from './bop';
import { RtuPartState, MonitoringState, RtuMaker, Frequency, FiberState } from './ft-enums';
import { NetAddress } from './net-address';
import { Trace } from './trace';

export class Rtu {
  rtuId!: string;
  nodeId!: string;
  rtuMaker!: RtuMaker;
  title!: string;

  mfid!: string | null;
  mfsn!: string | null;
  omid!: string | null;
  omsn!: string | null;
  serial!: string | null;
  version!: string | null;
  version2!: string | null;

  ownPortCount!: number;
  fullPortCount!: number;

  mainChannel!: NetAddress;
  mainChannelState!: RtuPartState;
  reserveChannel!: NetAddress;
  reserveChannelState!: RtuPartState;
  isReserveChannelSet!: boolean;
  otdrNetAddress!: NetAddress;
  monitoringMode!: MonitoringState;
  preciseMeas!: Frequency;
  preciseSave!: Frequency;
  fastSave!: Frequency;

  bops!: Bop[];
  traces!: Trace[];

  acceptableParams!: TreeOfAcceptableMeasurementParameters;

  // initialized during mapping
  isRtuAvailable!: boolean;
  isMonitoringOn!: boolean;
  bopsState!: RtuPartState;
  ownAttachedTraceCount!: number;
  fullAttachedTraceCount!: number;

  children!: any[];
  comment!: string | null;
}
